File: app/src/main/java/org/akvo/flow/presentation/navigation/FlowNavigationView.java
/*
 * Copyright (C) 2017-2019 Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Flow.
 *
 * Akvo Flow is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Flow is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Flow.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package org.akvo.flow.presentation.navigation;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.TextView;

import com.google.android.material.internal.NavigationMenuView;
import com.google.android.material.navigation.NavigationView;

import org.akvo.flow.R;
import org.akvo.flow.app.FlowApp;
import org.akvo.flow.domain.SurveyGroup;
import org.akvo.flow.injector.component.ApplicationComponent;
import org.akvo.flow.injector.component.DaggerViewComponent;
import org.akvo.flow.injector.component.ViewComponent;
import org.akvo.flow.presentation.SnackBarManager;

import java.util.List;

import javax.inject.Inject;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class FlowNavigationView extends NavigationView implements IFlowNavigationView {

    private TextView currentUserTextView;
    private TextView surveyTitleTextView;
    private RecyclerView surveysRecyclerView;
    private RecyclerView usersRecyclerView;
    private DrawerNavigationListener drawerNavigationListener;
    private SurveyAdapter surveyAdapter;
    private UserAdapter usersAdapter;
    private Drawable hideUsersDrawable;
    private Drawable showUsersDrawable;
    private View userHeader;

    @Inject
    FlowNavigationPresenter presenter;

    @Inject
    SnackBarManager snackBarManager;

    public FlowNavigationView(Context context) {
        this(context, null);
    }

    public FlowNavigationView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public FlowNavigationView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    private void init() {
        initialiseInjector();
        presenter.setView(this);
        getViewTreeObserver().addOnGlobalLayoutListener(
                new ViewTreeObserver.OnGlobalLayoutListener() {
                    @Override
                    public void onGlobalLayout() {
                        getViewTreeObserver().removeGlobalOnLayoutListener(this);
                        initViews();
                        setNavigationItemListener();
                        initCurrentUserText();
                        initUserList();
                        initSurveyList();
                        presenter.load();
                    }
                });
    }

    private void initViews() {
        currentUserTextView = findViewById(R.id.current_user_name);
        surveyTitleTextView = findViewById(R.id.surveys_title_tv);
        surveysRecyclerView = findViewById(R.id.surveys_rv);
        usersRecyclerView = findViewById( R.id.users_rv);
        userHeader = findViewById(R.id.user_header);
        NavigationMenuView navigationMenuView = (NavigationMenuView) getChildAt(0);
        if (navigationMenuView != null) {
            navigationMenuView.setVerticalScrollBarEnabled(false);
        }
    }

    private void initialiseInjector() {
        ViewComponent viewComponent =
                DaggerViewComponent.builder().applicationComponent(getApplicationComponent())
                        .build();
        viewComponent.inject(this);
    }

    private ApplicationComponent getApplicationComponent() {
        return ((FlowApp) getContext().getApplicationContext()).getApplicationComponent();
    }

    private void initUserList() {
        final Context context = getContext();
        usersRecyclerView.setLayoutManager(new LinearLayoutManager(context));
        usersAdapter = new UserAdapter(context);
        usersRecyclerView.setAdapter(usersAdapter);
        usersRecyclerView.addOnItemTouchListener(new RecyclerItemClickListener(context,
                new RecyclerItemClickListener.OnItemClickListener() {
                    @Override
                    public void onItemClick(View childView, int position) {
                        presenter.onUserSelected(usersAdapter.getItem(position));
                    }

                    @Override
                    public void onItemLongPress(View childView, int position) {
                        presenter.onUserLongPress(usersAdapter.getItem(position));
                    }
                })
        );
    }

    private void setNavigationItemListener() {
        findViewById(R.id.settings_tv).setOnClickListener(v -> {
            if (drawerNavigationListener != null) {
                drawerNavigationListener.navigateToSettings();
            }
        });

        findViewById(R.id.help_tv).setOnClickListener(v -> {
            if (drawerNavigationListener != null) {
                drawerNavigationListener.navigateToHelp();
            }
        });

        findViewById(R.id.about_tv).setOnClickListener(v -> {
            if (drawerNavigationListener != null) {
                drawerNavigationListener.navigateToAbout();
            }
        });
        findViewById(R.id.offline_maps_tv).setOnClickListener(v -> {
            drawerNavigationListener.navigateToOfflineMaps();
        });
    }

    private void initSurveyList() {
        final Context context = getContext();
        surveysRecyclerView.setLayoutManager(new LinearLayoutManager(context));
        surveyAdapter = new SurveyAdapter(context);
        surveysRecyclerView.setAdapter(surveyAdapter);
        surveysRecyclerView.addOnItemTouchListener(new RecyclerItemClickListener(context,
                new RecyclerItemClickListener.OnItemClickListener() {
                    @Override
                    public void onItemClick(View childView, int position) {
                        onSurveyItemTap(position);
                    }

                    @Override
                    public void onItemLongPress(View childView, int position) {
                        onSurveyItemLongPress(position, surveyAdapter);
                    }
                })
        );
    }

    private void initCurrentUserText() {
        hideUsersDrawable = ContextCompat.getDrawable(getContext(), R.drawable.ic_expand_less);
        showUsersDrawable = ContextCompat.getDrawable(getContext(), R.drawable.ic_expand_more);
        userHeader.setOnClickListener(v -> {
            if (surveysRecyclerView.getVisibility() == VISIBLE) {
                updateTextViewDrawable(hideUsersDrawable);
                surveyTitleTextView.setVisibility(GONE);
                surveysRecyclerView.setVisibility(GONE);
                usersRecyclerView.setVisibility(VISIBLE);
            } else {
                updateTextViewDrawable(showUsersDrawable);
                surveyTitleTextView.setVisibility(VISIBLE);
                surveysRecyclerView.setVisibility(VISIBLE);
                usersRecyclerView.setVisibility(GONE);

            }
        });
        userHeader.setOnLongClickListener(v -> {
            presenter.onCurrentUserLongPress();
            return true;
        });
    }

    private void updateTextViewDrawable(Drawable drawable) {
        currentUserTextView.setCompoundDrawablesWithIntrinsicBounds(null, null, drawable, null);
    }

    private void onSurveyItemLongPress(int position, SurveyAdapter adapter) {
        ViewSurvey viewSurvey = adapter.getItem(position);
        if (viewSurvey != null) {
            DialogFragment dialogFragment = SurveyDeleteConfirmationDialog
                    .newInstance(viewSurvey);
            dialogFragment.show(getSupportFragmentManager(), SurveyDeleteConfirmationDialog.TAG);
        }
    }

    private void onSurveyItemTap(int position) {
        presenter.onSurveyItemTap(surveyAdapter.getItem(position));
    }

    public void setDrawerNavigationListener(DrawerNavigationListener drawerNavigationListener) {
        this.drawerNavigationListener = drawerNavigationListener;
    }

    @Override
    public void displaySurveys(List<ViewSurvey> surveys, Long selectedSurveyId) {
        surveyAdapter.setSurveys(surveys, selectedSurveyId);
    }

    @Override
    public void notifySurveyDeleted(long surveyGroupId) {
        if (drawerNavigationListener != null) {
            drawerNavigationListener.onSurveyDeleted(surveyGroupId);
        }
    }

    @Override
    public void selectSurvey(SurveyGroup surveyGroup) {
        if (drawerNavigationListener != null) {
            drawerNavigationListener.onSurveySelected(surveyGroup);
            surveyAdapter.updateSelected(surveyGroup.getId());
        }
    }

    @Override
    public void displayUsers(String selectedUserName, List<ViewUser> viewUsers) {
        usersAdapter.setUsers(viewUsers);
        currentUserTextView.setText(selectedUserName);
    }

    @Override
    public void displayAddUser() {
        CreateUserDialog dialog = new CreateUserDialog();
        dialog.show(getSupportFragmentManager(), CreateUserDialog.TAG);
    }

    @Override
    public void displaySurveyError() {
        snackBarManager.displaySnackBar(this, R.string.surveys_error, getContext());
    }

    @Override
    public void displayUsersError() {
        snackBarManager.displaySnackBar(this, R.string.users_error, getContext());
    }

    @Override
    public void displayErrorDeleteSurvey() {
        snackBarManager.displaySnackBar(this, R.string.survey_delete_error, getContext());
    }

    @Override
    public void displayErrorSelectSurvey() {
        snackBarManager.displaySnackBar(this, R.string.survey_select_error, getContext());
    }

    @Override
    public void displayUserEditError() {
        snackBarManager.displaySnackBar(this, R.string.user_edit_error, getContext());
    }

    @Override
    public void displayUserDeleteError() {
        snackBarManager.displaySnackBar(this, R.string.user_delete_error, getContext());
    }

    @Override
    public void displayUserSelectError() {
        snackBarManager.displaySnackBar(this, R.string.user_select_error, getContext());
    }

    @Override
    public void displayEditUser(ViewUser currentUser) {
        DialogFragment fragment = EditUserDialog.newInstance(currentUser);
        fragment.show(getSupportFragmentManager(), EditUserDialog.TAG);
    }

    @Override
    public void onUserLongPress(ViewUser viewUser) {
        DialogFragment dialogFragment = UserOptionsDialog.newInstance(viewUser);
        dialogFragment.show(getSupportFragmentManager(), UserOptionsDialog.TAG);
    }

    private FragmentManager getSupportFragmentManager() {
        return ((AppCompatActivity) getContext()).getSupportFragmentManager();
    }

    public void onSurveyDeleteConfirmed(long surveyGroupId) {
        presenter.onDeleteSurvey(surveyGroupId);
    }

    @SuppressLint("RestrictedApi") @Override
    protected void onDetachedFromWindow() {
        presenter.destroy();
        super.onDetachedFromWindow();
    }

    public void editUser(ViewUser viewUser) {
        presenter.editUser(viewUser);
    }

    public void deleteUser(ViewUser viewUser) {
        presenter.deleteUser(viewUser);
    }

    public void createUser(String userName) {
        presenter.createUser(userName);
    }

    public interface DrawerNavigationListener {

        void onSurveySelected(SurveyGroup surveyGroup);

        void onSurveyDeleted(long surveyGroupId);

        void navigateToHelp();

        void navigateToAbout();

        void navigateToSettings();

        void navigateToOfflineMaps();
    }
}

File: app/src/main/java/org/akvo/flow/presentation/navigation/FlowNavigationView.kt
/*
 * Copyright (C) 2017-2020 Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Flow.
 *
 * Akvo Flow is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Flow is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Flow.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package org.akvo.flow.presentation.navigation

import android.content.Context
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.internal.NavigationMenuView
import com.google.android.material.navigation.NavigationView
import org.akvo.flow.R
import org.akvo.flow.app.FlowApp
import org.akvo.flow.utils.entity.SurveyGroup
import org.akvo.flow.injector.component.ApplicationComponent
import org.akvo.flow.injector.component.DaggerViewComponent
import org.akvo.flow.uicomponents.SnackBarManager
import javax.inject.Inject

class FlowNavigationView @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : NavigationView(
    context!!, attrs, defStyleAttr
), IFlowNavigationView {
    private lateinit var currentUserTextView: TextView
    private lateinit var surveyTitleTextView: TextView
    private lateinit var surveysRecyclerView: RecyclerView
    private lateinit var usersRecyclerView: RecyclerView
    private lateinit var surveyAdapter: SurveyAdapter
    private lateinit var usersAdapter: UserAdapter
    private lateinit var userHeader: View

    private var hideUsersDrawable: Drawable? = null
    private var showUsersDrawable: Drawable? = null
    private var drawerNavigationListener: DrawerNavigationListener? = null

    @Inject
    lateinit var presenter: FlowNavigationPresenter

    @Inject
    lateinit var snackBarManager: SnackBarManager

    init {
        init()
    }

    private fun init() {
        initialiseInjector()
        presenter.setView(this)
        viewTreeObserver.addOnGlobalLayoutListener(
            object : OnGlobalLayoutListener {
                override fun onGlobalLayout() {
                    viewTreeObserver.removeOnGlobalLayoutListener(this)
                    initViews()
                    setNavigationItemListener()
                    initCurrentUserText()
                    initUserList()
                    initSurveyList()
                    presenter.load()
                }
            })
    }

    private fun initViews() {
        currentUserTextView = findViewById(R.id.current_user_name)
        surveyTitleTextView = findViewById(R.id.surveys_title_tv)
        surveysRecyclerView = findViewById(R.id.surveys_rv)
        usersRecyclerView = findViewById(R.id.users_rv)
        userHeader = findViewById(R.id.user_header)
        val navigationMenuView = getChildAt(0) as NavigationMenuView
        navigationMenuView.isVerticalScrollBarEnabled = false
    }

    private fun initialiseInjector() {
        val viewComponent = DaggerViewComponent.builder().applicationComponent(
            applicationComponent
        )
            .build()
        viewComponent.inject(this)
    }

    private val applicationComponent: ApplicationComponent
        get() = (context.applicationContext as FlowApp).getApplicationComponent()

    private fun initUserList() {
        val context = context
        usersRecyclerView.layoutManager = LinearLayoutManager(context)
        usersAdapter = UserAdapter(context)
        usersRecyclerView.adapter = usersAdapter
        usersRecyclerView.addOnItemTouchListener(
            RecyclerItemClickListener(context,
                object : RecyclerItemClickListener.OnItemClickListener {
                    override fun onItemClick(childView: View, position: Int) {
                        presenter.onUserSelected(usersAdapter.getItem(position))
                    }

                    override fun onItemLongPress(childView: View, position: Int) {
                        presenter.onUserLongPress(usersAdapter.getItem(position))
                    }
                })
        )
    }

    private fun setNavigationItemListener() {
        findViewById<View>(R.id.settings_tv).setOnClickListener {
            drawerNavigationListener?.navigateToSettings()
        }
        findViewById<View>(R.id.help_tv).setOnClickListener {
            drawerNavigationListener?.navigateToHelp()
        }
        findViewById<View>(R.id.about_tv).setOnClickListener {
            drawerNavigationListener?.navigateToAbout()
        }
        findViewById<View>(R.id.offline_maps_tv).setOnClickListener {
            drawerNavigationListener?.navigateToOfflineMaps()
        }
    }

    private fun initSurveyList() {
        val context = context
        surveysRecyclerView.layoutManager = LinearLayoutManager(context)
        surveyAdapter = SurveyAdapter(context)
        surveysRecyclerView.adapter = surveyAdapter
        surveysRecyclerView.addOnItemTouchListener(
            RecyclerItemClickListener(context,
                object : RecyclerItemClickListener.OnItemClickListener {
                    override fun onItemClick(childView: View, position: Int) {
                        onSurveyItemTap(position)
                    }

                    override fun onItemLongPress(childView: View, position: Int) {
                        onSurveyItemLongPress(position, surveyAdapter)
                    }
                })
        )
    }

    private fun initCurrentUserText() {
        hideUsersDrawable = ContextCompat.getDrawable(context, R.drawable.ic_expand_less)
        showUsersDrawable = ContextCompat.getDrawable(context, R.drawable.ic_expand_more)
        userHeader.setOnClickListener {
            if (surveysRecyclerView.visibility == VISIBLE) {
                updateTextViewDrawable(hideUsersDrawable)
                surveyTitleTextView.visibility = GONE
                surveysRecyclerView.visibility = GONE
                usersRecyclerView.visibility = VISIBLE
            } else {
                updateTextViewDrawable(showUsersDrawable)
                surveyTitleTextView.visibility = VISIBLE
                surveysRecyclerView.visibility = VISIBLE
                usersRecyclerView.visibility = GONE
            }
        }
        userHeader.setOnLongClickListener {
            presenter.onCurrentUserLongPress()
            true
        }
    }

    private fun updateTextViewDrawable(drawable: Drawable?) {
        currentUserTextView.setCompoundDrawablesWithIntrinsicBounds(null, null, drawable, null)
    }

    private fun onSurveyItemLongPress(position: Int, adapter: SurveyAdapter) {
        val viewSurvey = adapter.getItem(position)
        val dialogFragment: DialogFragment = SurveyDeleteConfirmationDialog
            .newInstance(viewSurvey)
        dialogFragment.show(supportFragmentManager, SurveyDeleteConfirmationDialog.TAG)
    }

    private fun onSurveyItemTap(position: Int) {
        presenter.onSurveyItemTap(surveyAdapter.getItem(position))
    }

    fun setDrawerNavigationListener(drawerNavigationListener: DrawerNavigationListener?) {
        this.drawerNavigationListener = drawerNavigationListener
    }

    override fun displaySurveys(surveys: List<ViewSurvey>, selectedSurveyId: Long?) {
        surveyAdapter.setSurveys(surveys, selectedSurveyId!!)
        val newSurveys = surveys.filter {
            !it.viewed
        }
        drawerNavigationListener?.updateDrawerIcon(newSurveys)
    }

    override fun notifySurveyDeleted(surveyGroupId: Long) {
            drawerNavigationListener?.onSurveyDeleted(surveyGroupId)
    }

    override fun selectSurvey(surveyGroup: SurveyGroup?) {
        drawerNavigationListener?.let { drawerNavigationListener ->
            drawerNavigationListener.onSurveySelected(surveyGroup)
            surveyAdapter.updateSelected(surveyGroup!!.id)
        }
    }

    override fun displayUsers(selectedUserName: String?, viewUsers: List<ViewUser>?) {
        usersAdapter.setUsers(viewUsers)
        currentUserTextView.text = selectedUserName
    }

    override fun displayAddUser() {
        val dialog = CreateUserDialog()
        dialog.show(supportFragmentManager, CreateUserDialog.TAG)
    }

    override fun displaySurveyError() {
        snackBarManager.displaySnackBar(this, R.string.surveys_error, context)
    }

    override fun displayUsersError() {
        snackBarManager.displaySnackBar(this, R.string.users_error, context)
    }

    override fun displayErrorDeleteSurvey() {
        snackBarManager.displaySnackBar(this, R.string.survey_delete_error, context)
    }

    override fun displayErrorSelectSurvey() {
        snackBarManager.displaySnackBar(this, R.string.survey_select_error, context)
    }

    override fun displayUserEditError() {
        snackBarManager.displaySnackBar(this, R.string.user_edit_error, context)
    }

    override fun displayUserDeleteError() {
        snackBarManager.displaySnackBar(this, R.string.user_delete_error, context)
    }

    override fun displayUserSelectError() {
        snackBarManager.displaySnackBar(this, R.string.user_select_error, context)
    }

    override fun displayEditUser(currentUser: ViewUser?) {
        val fragment: DialogFragment = EditUserDialog.newInstance(currentUser)
        fragment.show(supportFragmentManager, EditUserDialog.TAG)
    }

    override fun onUserLongPress(viewUser: ViewUser?) {
        val dialogFragment: DialogFragment = UserOptionsDialog.newInstance(viewUser)
        dialogFragment.show(supportFragmentManager, UserOptionsDialog.TAG)
    }

    private val supportFragmentManager: FragmentManager
        get() = (context as AppCompatActivity).supportFragmentManager

    fun onSurveyDeleteConfirmed(surveyGroupId: Long) {
        presenter.onDeleteSurvey(surveyGroupId)
    }

    override fun onDetachedFromWindow() {
        presenter.destroy()
        super.onDetachedFromWindow()
    }

    fun editUser(viewUser: ViewUser?) {
        presenter.editUser(viewUser)
    }

    fun deleteUser(viewUser: ViewUser?) {
        presenter.deleteUser(viewUser)
    }

    fun createUser(userName: String?) {
        presenter.createUser(userName)
    }

    interface DrawerNavigationListener {
        fun onSurveySelected(surveyGroup: SurveyGroup?)
        fun onSurveyDeleted(surveyGroupId: Long)
        fun navigateToHelp()
        fun navigateToAbout()
        fun navigateToSettings()
        fun navigateToOfflineMaps()
        fun updateDrawerIcon(newSurveys: List<ViewSurvey>)
    }
}
